packages/shared/client/ReplayClient.ts

```diff
@@ -241,11 +241,11 @@ export class ReplayClient implements ReplayClientInterface {
   }> {
     const sessionId = this.getSessionIdThrows();
 
-    // Don't try to fetch messages in unloaded regions.
-    // The result might be invalid (and may get cached by a Suspense caller).
-    await this._waitForRangeToBeLoaded(focusRange);
```

Comment:
Why was this okay to do?

If there is no focus range, and we try to fetch messages for the whole recording, we might end up asking about unloaded regions still, right?

`_waitForRangeToBeLoaded` is smart enough to handle this case.

https://github.com/replayio/devtools/blob/fcae274fba899c2ec5db8fbf12f0fe393400ab53/packages/shared/client/ReplayClient.ts#L976-L989

Reply:
Ah, sorry, I mentioned this in the linked loom explainer, but I did not write it in the commit message. We have two different methods for fetching console messages: `findMessages` and `findMessagesInRange`. Now, I've gone on this rant probably a dozen times at this point, but the reason that this *one* basic data resource has multiple methods for fetching exactly the same kind of information is because it has an `overflow` flag, so discovering console messages in a particular region might allow you to bypass that flag. Unfortunately, they have different requirements because of how they work.

`findMessages` is much closer to all of the other basic data resources: it does not interact with loaded or indexed regions at all. It's always available, and always returns all data for the whole recording, regardless of what is currently loaded. For instance, we do not wait for the whole recording to be loaded before calling `findSources`, nor do we expect that `findSources` will only return sources that were used or first downloaded in the currently loaded regions.

Because `findMessagesInRange` is designed to get around the overflow flag, it has to go directly to the running linker and ask it to run through particular part of the recording and re-collect the console messages. This requires a running linker in those regions, and so it requires that those regions be loaded. I think we should either make this endpoint return the actual window for which it ran with the results, or have it error if it can't run through the requested window, rather than what it does now (silently return a different window) but that's a slightly tangential topic.

Anyways, because of the way `findMessages` works, we can issue that request as soon as the Replay is loaded, and in *most* cases we will never need to request again, because there will be fewer than 1,000 console messages.

Reply:
Thank you for elaborating! That was very helpful. Could we add some abbreviated form of that as a comment inline for future me (or others) who might see this and think it's a mistake?

Reply:
Yeah sounds great, will do.

Reply:
I guess I'll add one other thing while I'm in here: it's really not the focus window that should be deciding which of these we use, but rather we should probably always try `findMessages` first and only in the case of overflow fall back to `findMessagesInRange` as the focus window changes. Otherwise, users who load a replay with a specific window encoded in the URL will still end up waiting for the regions to load before fetching the console messages (even though in most cases they could issue a single `findMessages` and be done). But that's a larger change I think.

Reply:
I think it's more complicated than the above paragraph makes it sound– because in a bad case, we'd end up fetching all messages, seeing overflow, and then re-fetching a slice of messages. Adding the extra logic for that doesn't seem worth it to me.